near-sdk/src/store/free_list/mod.rs

```diff
@@ -301,6 +306,44 @@ mod tests {
         assert_eq!(bucket.get(i3), Some(&4));
     }
 
+    #[test]
+    fn defrag() {
+        let mut bucket = FreeList::new(b"b");
+        assert!(bucket.is_empty());
+        let i1 = bucket.insert(1u8);
+        let i2 = bucket.insert(2u8);
+        let i3 = bucket.insert(3u8);
+        let i4 = bucket.insert(4u8);
+        let i5 = bucket.insert(5u8);
+        let i6 = bucket.insert(6u8);
+        let i7 = bucket.insert(7u8);
+        let i8 = bucket.insert(8u8);
+        assert_eq!(bucket.occupied_count, 8);
+
+        //Empty, Empty, Empty, Empty, Occupied, Occupied, Empty, Empty
+        bucket.remove(i2);
+        bucket.remove(i4);
+        bucket.remove(i1);
+        bucket.remove(i7);
+        bucket.remove(i3);
+        bucket.remove(i8);
+        assert_eq!(bucket.occupied_count, 2);
+
+        let free_items = navigate_empty_slots(&bucket);
+        assert_eq!(free_items, 6);
```

Comment:
so one note about this implementation, this should actually just clear the last elements from storage altogether (and shrink the underlying Vector, set the `first_free` back to None). This way storage costs can be refunded on this operation

Reply:
I see! This should make the logic simpler, I guess.

Reply:
yeah, I would suggest using `drain(<num of elements>..)` and then iterating through to pull filled cells to move into the vacant entries, but up to you and what you had planned to implement!

Reply:
If the free slots chain is in drained vector then it becomes difficult to traverse through them. I just pushed a PR. But, now, I think the better option is to 
1. Traverse through free slots chain and store the free slots that needs replacement (`self.occupied_count`) in a `Vec`
2. Pop items from the `self.elements` and if its occupied, then place it in one of the free slots in the `Vec`

Let me know if you are okay with above approach?